### PR TITLE
Clarify syscall ABI. Clear caller saved regs.

### LINKED ISSUE #2: Kernel leaks information during syscalls, in caller-save registers
Perhaps the kernel should trash them before returning.

### PR DESCRIPTION
**Pull request by allanlw**
This fixes #2

### CODE REVIEW

File: kernel/isr-asm.s [hunk 1]
```diff
@@ -29,22 +29,67 @@
...
+	/* Translate syscall arguments into appropriat regs for handler */
+	mov %rbx,%rdi
+	mov %rcx,%rsi
 	call *%rax
+
+	/* Clear registers to prevent information leakage from kernel mode
+	 * to usermode. For *callee* saved registers, it would be an ABI
+	 * to not have preserved them, so we only need to clobber caller
+	 * saved registers */
+	clear_caller_saved_reg_except_rax
```
Comment (gsilvis) [inline kernel/isr-asm.s]: Do any syscalls not have a return value?  If there are any, then rax needs to be cleared for them.

Comment (gsilvis) [inline kernel/isr-asm.s]: If there aren't any, we should document that doing that is unsafe, and recommend returning a code.

Comment (allanlw) [inline kernel/isr-asm.s]: I'm not sure how to fix that. That really should be checked by the type system, but the syscall dispatch table is in ASM. How could you tell the difference..?

Comment (gsilvis) [inline kernel/isr-asm.s]: Yeah, that's why I think we should just forbid it

**Comment by gsilvis** [comment]
Ugh, this is really persnickety, but we technically need to make sure we're not leaking information in the top half of %rax, which won't necessarily get written to when returning an int.

**Comment by allanlw** [comment]
Yeah I considered that but don't have a good answer.

Maybe there should really be a series of trampoline functions for wrapping syscalls in?
Also it would be *really* nice to have a way to programmatically catch regressions here.

**Pull request merged**